packages/client/hmi-client/src/services/workflow.ts

```diff
@@ -331,6 +331,20 @@ export class WorkflowRegistry {
 	}
 }
 
+export function cascadeInvalidateDownstream(
+	wf: Workflow,
+	sourceNode: WorkflowNode<any>,
+	nodesCache: Map<string, WorkflowNode<any>>
+) {
+	const outgoingEdges = wf.edges.filter((edge) => edge.source === sourceNode.id);
+	outgoingEdges.forEach((edge) => {
+		const targetNode = nodesCache.get(edge.target);
+		if (!targetNode) return;
+		targetNode.status = OperatorStatus.INVALID;
+		cascadeInvalidateDownstream(wf, targetNode, nodesCache);
+	});
+}
```

Comment:
`Filter` is actually worse than `find` because you must traverse the entire array 😄 

My idea of is that you build `id => node[]` map/relationship ahead of time so you avoid doing any lookups or filters, can we do that?